src/artm/core/cooccurrence_collector.cc

```diff
@@ -0,0 +1,1053 @@
+            ++num_of_documents_token_occurred_in_[first_token_id];
+          }
+          // 5.d) Take windows_width tokens (parameter) to the right of the current one
+          // If there are some words beginnig on '|' in the text the window should be extended
+          // and it's extended using not_a_word_counter
+          std::string second_token_modality = first_token_modality;
+          unsigned not_a_word_counter = 0;
+          // Loop through tokens in the window
+          for (unsigned k = 1; k <= config_.cooc_window_width() + not_a_word_counter && j + k < doc.size(); ++k) {
+            if (doc[j + k].empty()) {
+              continue;
+            }
+            if (doc[j + k][0] == '|') {
+              second_token_modality = doc[j + k];
+              ++not_a_word_counter;
+              continue;
+            }
+            if (first_token_modality != second_token_modality) {
+              continue;
+            }
+            int second_token_id = vocab_.FindTokenId(doc[j + k], second_token_modality);
+            if (second_token_id == TOKEN_NOT_FOUND) {
+              continue;
+            }
+            // 5.e) When it's known these 2 tokens are valid, remember their co-occurrence
+            // Here portion.size() is used to identify a document (it's unique id during one portion of documents)
+            if (config_.use_symetric_cooc()) {
+              if (first_token_id < second_token_id) {
+                cooc_stat_holder.SavePairOfTokens(first_token_id, second_token_id, portion.size());
+              } else if (first_token_id > second_token_id) {
+                cooc_stat_holder.SavePairOfTokens(second_token_id, first_token_id, portion.size());
+              } else {
+                cooc_stat_holder.SavePairOfTokens(first_token_id, first_token_id, portion.size(), 2);
+              }
+            } else {
+              cooc_stat_holder.SavePairOfTokens(first_token_id, second_token_id, portion.size());
+              cooc_stat_holder.SavePairOfTokens(second_token_id, first_token_id, portion.size());
+            }
+            local_num_of_pairs += 2;  // statistics for ppmi
+          }
+        }
+      }
+      if (!cooc_stat_holder.storage_.empty()) {
+        // This function saves gathered statistics on disk
+        // After saving on disk statistics from all the batches needs to be merged
+        // This is implemented in ReadAndMergeCooccurrenceBatches(), so the next step is to call this method
+        // Sorting is needed before storing all pairs of tokens on disk (it's for future agregation)
+        UploadOnDisk(cooc_stat_holder);
+      }
+    }
+    {
+      std::unique_lock<std::mutex> lock(total_num_of_pairs_arg_access);
+      total_num_of_pairs_ += local_num_of_pairs;
+    }
+  };
+  // Launch reading and storing pairs of tokens in parallel
+  std::vector<std::shared_future<void>> tasks;
+  for (int i = 0; i < config_.num_of_cpu(); ++i) {
```

Comment:
can't it be done in one loop?

Reply:
no, then it wouldn't be multithread. It'll just wait for task to finsh and will launch another